pkg/kubelet/kubelet_node_status_test.go

```diff
@@ -896,3 +897,193 @@ func TestRegisterExistingNodeWithApiserver(t *testing.T) {
+			newNode:         newNode(false, "a"),
+			createError:     alreadyExists,
+			getError:        conflict,
+			expectedResult:  false,
+			expectedActions: 2,
+		},
+		{
+			name:            "update existing node failed",
+			newNode:         newNode(false, "a"),
+			createError:     alreadyExists,
+			existingNode:    newNode(true, "a"),
+			updateError:     conflict,
+			expectedResult:  false,
+			expectedActions: 3,
+		},
+		{
+			name:            "delete existing node failed",
+			newNode:         newNode(false, "b"),
+			createError:     alreadyExists,
+			existingNode:    newNode(false, "a"),
+			deleteError:     conflict,
+			expectedResult:  false,
+			expectedActions: 3,
+		},
+	}
+
+	for _, tc := range cases {
+		testKubelet := newTestKubelet(t, false /* controllerAttachDetachEnabled is a don't-care for this test */)
+		kubelet := testKubelet.kubelet
+		kubeClient := testKubelet.fakeKubeClient
+
+		kubeClient.AddReactor("create", "nodes", func(action core.Action) (bool, runtime.Object, error) {
+			return true, nil, tc.createError
+		})
+		kubeClient.AddReactor("get", "nodes", func(action core.Action) (bool, runtime.Object, error) {
+			// Return an existing (matching) node on get.
+			return true, tc.existingNode, tc.getError
+		})
+		kubeClient.AddReactor("update", "nodes", func(action core.Action) (bool, runtime.Object, error) {
+			return true, nil, tc.updateError
+		})
+		kubeClient.AddReactor("delete", "nodes", func(action core.Action) (bool, runtime.Object, error) {
+			return true, nil, tc.deleteError
+		})
+		kubeClient.AddReactor("*", "*", func(action core.Action) (bool, runtime.Object, error) {
+			return true, nil, fmt.Errorf("no reaction implemented for %s", action)
+		})
+
+		result := kubelet.registerWithApiServerAttempt(tc.newNode)
+		if e, a := tc.expectedResult, result; e != a {
+			t.Errorf("%v: unexpected result; expected %v got %v", tc.name, e, a)
+			continue
+		}
+
+		actions := kubeClient.Actions()
+		if e, a := tc.expectedActions, len(actions); e != a {
+			t.Errorf("%v: unexpected number of actions, expected %v, got %v", tc.name, e, a)
+		}
+
+		if tc.testSavedNode {
+			var savedNode *api.Node
+			var ok bool
+
+			t.Logf("actions: %v: %+v", len(actions), actions)
+			action := actions[tc.savedNodeIndex]
+			if action.GetVerb() == "create" {
+				createAction, _ := action.(core.CreateAction)
+				savedNode, ok = createAction.GetObject().(*api.Node)
+				if !ok {
+					t.Errorf("%v: unexpected type; couldn't convert to *api.Node: %+v", tc.name, createAction.GetObject())
+					continue
+				}
+			} else if action.GetVerb() == "update" {
+				updateAction, _ := action.(core.UpdateAction)
+				savedNode, ok = updateAction.GetObject().(*api.Node)
+				if !ok {
+					t.Errorf("%v: unexpected type; couldn't convert to *api.Node: %+v", tc.name, updateAction.GetObject())
+					continue
+				}
+			}
+
+			annValue, ok := savedNode.Annotations[volumehelper.ControllerManagedAttachAnnotation]
+			if tc.savedNodeCMAD {
```

Comment:
Could you simplify this to just compare tc.savedNodeCMAD to annValue? annValue will be "" if it doesn't exist in the map, and if it does exist, then a direct comparison should be sufficient, right?